### PR TITLE
add Gather implementation

### LINKED ISSUE #19957: DNN/ONNX: Gather layer with 4 indices is not supported
Hi,
I use opencv 4.5.2 on my ubuntu 20.04 desktop, when I load the onnx model provided by onnx officially, for example yolov4（https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/yolov4) :
```
[ERROR:0] global /home/jim/opencv/opencv4.5.2/opencv-4.5.2/modules/dnn/src/onnx/onnx_importer.cpp (2108) handleNode DNN/ONNX: ERROR during processing node with 1 inputs and 1 outputs: [Log]:(StatefulPartitionedCall/model/lambda/Log:0)
Traceback (most recent call last):
  File "/home/jim/DL4CV_lesson/cv2 blobFromImage/read_onnx.py", line 16, in <module>
    model = cv2.dnn.readNetFromONNX('yolov4.onnx')
cv2.error: OpenCV(4.5.2) /home/jim/opencv/opencv4.5.2/opencv-4.5.2/modules/dnn/src/onnx/onnx_importer.cpp:2129: error: (-2:Unspecified error) in function 'handleNode'
> Node [Log]:(StatefulPartitionedCall/model/lambda/Log:0) parse error: OpenCV(4.5.2) /home/jim/opencv/opencv4.5.2/opencv-4.5.2/modules/dnn/src/dnn.cpp:621: error: (-2:Unspecified error) Can't create layer "StatefulPartitionedCall/model/lambda/Log:0" of type "Log" in function 'getLayerInstance'
> 
```
Thanks a lot !!!

### LINKED ISSUE #20439: Can't import ONNX model produced by keras2onnx
##### System information (version)
- OpenCV(4.5.3-dev)
- Operating System / Platform => Ubuntu 20.04
- Compiler => gcc 9.3.0

Description of the issue:
I am using the caption model which is converted using keras2onnx. I am unable to load the model by using readNet or readNetFromONNX in opencv. And I have installed Opencv from source (OpenCV(4.5.3-dev)).

# steps to reproduce : 
Code is as below : (C++)
```
std::string Caption_Model = "Caption_Model.onnx";
cv::dnn::Net net = cv::dnn::readNetFromONNX(Caption_Model);
```
Error message:
```
Error : Node [Gather]:(embedding_1/Identity:0) parse error: OpenCV(4.5.3-dev) /opencv_build/opencv/modules/dnn/src/onnx/onnx_importer.cpp:356: error: (-5:Bad argument) Blob casted not found in const blobs in function 'getBlob'
Aborted (core dumped)
```

You can find the architecture of the caption model in the screenshot.
Caption_Model

### PR DESCRIPTION
**Pull request by rogday**
**Merge with extra:** https://github.com/opencv/opencv_extra/pull/1007
Should resolve #20439, resolve #19957

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by zihaomu** [comment]
Hi @rogday, How is the PR progress so far? Maybe I can do something if you want.

**Comment by rogday** [comment]
Hi @zihaomu, I think we need to wait for merge of @fengyuentau's extra blobs info so that we can reuse it for 1-d and 0-d support here. Otherwise I think it's pretty much done.

**Comment by fengyuentau** [comment]
Hello @rogday , we came across [the problem of lacking support for gather op with multiple indices](https://github.com/opencv/opencv_zoo/pull/82#discussion_r950709754). I wonder if you have time working on this pull request. If no, me or someone on my side can help you finish this pull request.

**Comment by rogday** [comment]
> Hello @rogday , we came across [the problem of lacking support for gather op with multiple indices](https://github.com/opencv/opencv_zoo/pull/82#discussion_r950709754). I wonder if you have time working on this pull request. If no, me or someone on my side can help you finish this pull request.

Hi! Sorry for the delay. I'll give it a shot next week. Would you be okay with that?

**Comment by fengyuentau** [comment]
> I'll give it a shot next week. Would you be okay with that?

Sure, no problem. Thank you!

File: modules/dnn/test/test_onnx_importer.cpp [hunk 1]
```diff
@@ -205,13 +205,17 @@ TEST_P(Test_ONNX_layers, Convolution_variable_weight_bias)
 
 TEST_P(Test_ONNX_layers, Gather)
 {
-    if (backend == DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019 && target == DNN_TARGET_MYRIAD)
-        applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_MYRIAD, CV_TEST_TAG_DNN_SKIP_IE_NN_BUILDER);
```
Comment (alalek) [inline modules/dnn/test/test_onnx_importer.cpp:209]: Removal of 'skip" checks means that this test passes.
It is not true, so we should kept existed conditions.

Don't change existed, add new only.

File: modules/dnn/test/test_onnx_importer.cpp [hunk 2]
```diff
@@ -205,13 +205,17 @@ TEST_P(Test_ONNX_layers, Convolution_variable_weight_bias)
...
+    if (backend != DNN_BACKEND_OPENCV || target != DNN_TARGET_CPU)
+        throw SkipTestException("Only CPU is supported");  // FIXIT use tags
```
Comment (alalek) [inline modules/dnn/test/test_onnx_importer.cpp:209]: When do we lost OpenCV/OpenCL support?

File: modules/dnn/test/test_onnx_importer.cpp (hunk 1 above)
Comment (rogday) [reply modules/dnn/test/test_onnx_importer.cpp:209]: This PR replaces old implementation and doesn't support anything except for CPU for now. Subset of Gather used to be implemented via Slice + Reshape, which support multiple backends.

File: modules/dnn/test/test_onnx_importer.cpp (hunk 2 above)
Comment (rogday) [reply modules/dnn/test/test_onnx_importer.cpp:209]: See the comment above.

Comment (alalek) [reply modules/dnn/test/test_onnx_importer.cpp:209]: OpenCV/OpenCL code path should have automatic fallback implementation on CPU code.
Test should check that fallback works.

**Review by fengyuentau — APPROVED** [review]
Could we turn on some onnx conformance tests of gather? Others look good to me

**Comment by rogday** [comment]
> Could we turn on some onnx conformance tests of gather? Others look good to me

Unfortunately no - type(integer indices) and shape(less than 2d) issues in non-const context aren't handled by this PR.

File: modules/dnn/test/test_onnx_importer.cpp (hunk 2 above)
Comment (rogday) [reply modules/dnn/test/test_onnx_importer.cpp]: Updated the PR.

File: modules/dnn/src/init.cpp [hunk 3]
```diff
@@ -147,6 +147,7 @@ void initializeLayerFactory()
     CV_DNN_REGISTER_LAYER_CLASS(Const,          ConstLayer);
     CV_DNN_REGISTER_LAYER_CLASS(Arg,            ArgLayer);
     CV_DNN_REGISTER_LAYER_CLASS(Reciprocal,     ReciprocalLayer);
+    CV_DNN_REGISTER_LAYER_CLASS(Gather,            GatherLayer);
```
Comment (alalek) [inline modules/dnn/src/init.cpp:150]: Lets don't break alignment

**Review by alalek — APPROVED** [review]

File: modules/dnn/test/test_onnx_importer.cpp (hunk 2 above)
Comment (alalek) [reply modules/dnn/test/test_onnx_importer.cpp]: > Test should check that fallback works.

and OpenCL FP16 tests failed...

**Pull request merged**